### PR TITLE
Fix #3619

### LINKED ISSUE #3619: dune subst crashes when there is no commit
Hi,

This is a bit of an edge case but I think that's worth reporting.

When running `dune subst` in a git repository with no commit, the following error happens:

```cram
  $ cat > dune-project <<EOF
  > (lang dune 1.0)
  > (name foo)
  > EOF

  $ echo 'authors: [ "John Doe <john@doe.com>" ]' > foo.opam

  $ git init --quiet

  $ dune subst
           git (internal) (exit 128)
  /usr/bin/git describe --always --dirty > /tmp/dune.cram.43198erun.t/tmp/dune3a97db.output
  fatal: bad revision 'HEAD'
           git (internal) (exit 128)
  /usr/bin/git rev-parse HEAD > /tmp/dune.cram.43198erun.t/tmp/dune7e6051.output
  fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
  Use '--' to separate paths from revisions, like this:
  'git <command> [<revision>...] -- [<file>...]'
           git (internal) (exit 128)
  /usr/bin/git ls-tree -r --name-only HEAD > /tmp/dune.cram.43198erun.t/tmp/dune2547a1.output
  fatal: Not a valid object name HEAD
  [1]
```

This happened to me while opam pinning a new package. I started working on a fix, but it's a bit complex because we need to both get the git error code and its output, which we don't do at the moment, so I put that on hold.

### PR DESCRIPTION
**Pull request by rgrinberg**
When the repository is empty, substitute version with dummy values.

Fix #3619
cc @CraigFe and @emillon

### CODE REVIEW

File: src/dune_engine/vcs.ml [hunk 1]
```diff
@@ -114,25 +126,60 @@ let make_fun name ~output ~doc ~git ~hg =
...
   Staged.unstage
   @@ make_fun "vcs-describe"
        ~doc:"Obtain a nice description of the tip from the vcs"
-       ~output:(Simple (module String))
-       ~git:(fun t -> run t [ "describe"; "--always"; "--dirty" ])
-       ~hg:hg_describe
+       ~output:(Simple (module Option_output (String)))
+       ~git:(fun t ->
+         run_git t [ "describe"; "--always"; "--dirty" ]
+         |> Fiber.map ~f:(Option.map ~f:String.trim))
```
Comment (jeremiedimino) [inline src/dune_engine/vcs.ml:147]: `run_git` already trims the string

File: test/blackbox-tests/test-cases/github4429.t [hunk 2]
```diff
@@ -18,7 +18,6 @@ Create a repository with no HEAD commit:
 At the moment Dune fails, which is bad:
 
   $ dune exec ./main.exe 2>&1 | sed 's/.*\/git/{{ git }}/; s/> .*.output/> {{ output_file }}/g' 
-           git (internal) (exit 128)
-  {{ git }} describe --always --dirty > {{ output_file }}
+           git (internal)
   fatal: bad revision 'HEAD'
```
Comment (jeremiedimino) [inline test/blackbox-tests/test-cases/github4429.t:22]: Shouldn't we get rid of this message as well?

**Review by jeremiedimino — CHANGES_REQUESTED** [review]

File: test/blackbox-tests/test-cases/github4429.t (hunk 2 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/github4429.t:22]: Yeah, I suppose. I thought it would be nice to warn the user of this issue, but the error message is cryptic anyway.

File: src/dune_engine/vcs.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_engine/vcs.ml:147]: done.

File: test/blackbox-tests/test-cases/github4429.t (hunk 2 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/github4429.t:22]: I tried redirecting stderr to `/dev/null` but now I'm getting `# Error: dup2: : Bad file descriptor` everywhere when running `$ dune subst`. No idea why that is.

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/github4429.t:22]: is there a way to test if a git repo has no commit?

Comment (nojb) [reply test/blackbox-tests/test-cases/github4429.t:22]: What I use is `git rev-parse --verify HEAD`. If it succeeds you have a commit, otherwise, no commit.

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/github4429.t:22]: Yes, but it still errors with 128 and returns a cryptic message. So I'm not sure how it helps. It also seems annoying to run an extra command just for this edge case that occurs once a decade.

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/github4429.t:22]: In any case this dup2 error is not right.

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/github4429.t:22]: Can you reproduce this in the test suite with a `(ignore-stderr ..)`?

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/github4429.t]: Alright, I just made the redirect using a file.

**Comment by rgrinberg** [comment]
@jeremiedimino should be ready now

File: src/dune_engine/vcs.ml [hunk 3]
```diff
@@ -79,9 +79,22 @@ let run t args =
...
+      ~env:Env.initial
+      ~stderr_to:(Process.Io.file Config.dev_null Out)
+  in
+  let open Fiber.O in
+  let+ res = res in
+  match res with
+  | Ok s -> Some (String.trim s)
+  | Error 128 -> None
+  | Error n ->
+    User_error.raise [ Pp.textf "Unexpected response from `git` (code %d)" n ]
```
Comment (jeremiedimino) [inline src/dune_engine/vcs.ml:97]: I believe this is unreachable, i.e. `Process` will already raise in this case. So `assert false` might be enough

File: src/dune_engine/vcs.ml [hunk 4]
```diff
@@ -114,25 +127,58 @@ let make_fun name ~output ~doc ~git ~hg =
...
+    let open Fiber.O in
+    let+ res =
+      Process.run_capture_zero_separated (git_accept ()) (prog t) args
+        ~dir:t.root ~env:Env.initial
+    in
+    match res with
+    | Ok s -> s
+    | Error 128 -> []
+    | Error n ->
+      User_error.raise [ Pp.textf "Unexpected response from `git` (code %d)" n ]
```
Comment (jeremiedimino) [inline src/dune_engine/vcs.ml:177]: Same here

File: test/blackbox-tests/test-cases/github4429.t [hunk 5]
```diff
@@ -18,7 +18,4 @@ Create a repository with no HEAD commit:
 At the moment Dune fails, which is bad:
 
   $ dune exec ./main.exe 2>&1 | sed 's/.*\/git/{{ git }}/; s/> .*.output/> {{ output_file }}/g'
```
Comment (jeremiedimino) [inline test/blackbox-tests/test-cases/github4429.t:20]: We can simplify the invocation now:

```suggestion
  $ dune exec ./main.exe
```

**Review by jeremiedimino — APPROVED** [review]
Looks good. Just left a couple of suggestions.

**Comment by ejgallego** [comment]
@rgrinberg note the 2.8.6. milestone here, so we need to either cherry pick to 2.9, or move the milestone to 3.0

**Comment by ejgallego** [comment]
Moving this to 3.0 then, please retarget if you want to backport.
